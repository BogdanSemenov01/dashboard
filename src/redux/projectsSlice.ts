import { PayloadAction, createSlice } from '@reduxjs/toolkit'

type State = {
  projects: Array<Project>
  currentProjectId: number
}

export type Project = {
  title: string
  id: number
  tasks: Tasks

}


type Tasks = {
  queueTasks: StatusTasks
  developmentTasks: StatusTasks
  doneTasks: StatusTasks
}

type StatusTasks = Array<Task>


type Task = {
  id: number
  text: string
  description: string
  priority: 'low' | 'middle' | 'high'
  subTasks: Array<Subtask>
}

type Subtask = {
  id: number
  text: string
  isComplete: boolean
}


const initialState: State = {
  projects: [
    {
      title: 'DashBoard',
      id: 1,
      tasks: {
        queueTasks: [
          {
            id: 1,
            text: 'queue',
            description: '',
            priority: 'low',
            subTasks: [
              { id: 1, text: '1', isComplete: true },
              { id: 2, text: '2', isComplete: true },
              { id: 3, text: '3', isComplete: false },
              { id: 4, text: '4', isComplete: false },
            ]
          },
        ],
        developmentTasks: [
          {
            id: 2, 
            text: 'dev', 
            description: '', 
            priority: 'low', 
            subTasks: [
              { id: 1, text: 'dfdfdf', isComplete: false }
            ]
          },
        ],
        doneTasks: [
          { 
            id: 3, 
            text: 'done', 
            description: '', 
            priority: 'low',
            subTasks: [
              { id: 1, text: 'dfdfdf', isComplete: false }
            ]
           },
        ],
      }
    },
    {
      title: 'Portfolio',
      id: 2,
      tasks: {
        queueTasks: [
          { 
            id: 3, 
            text: 'done', 
            description: '', 
            priority: 'low',
            subTasks: [
              { id: 1, text: 'dfdfdf', isComplete: false }
            ]
           },
        ],
        developmentTasks: [
          { 
            id: 3, 
            text: 'done', 
            description: '', 
            priority: 'low',
            subTasks: [
              { id: 1, text: 'dfdfdf', isComplete: false }
            ]
           },
        ],
        doneTasks: [
          { 
            id: 3, 
            text: 'done', 
            description: '', 
            priority: 'low',
            subTasks: [
              { id: 1, text: 'dfdfdf', isComplete: false }
            ]
           },
        ],
      }
    }
  ],
  currentProjectId: 0,
}

export const projectsSlice = createSlice({
  name: 'projects',
  initialState,
  reducers: {
    createProject: (state, action: PayloadAction<{text: string}>) => {
      state.projects.push({
        title: action.payload.text,
        id: Date.now(),
        tasks: {
          queueTasks: [],
          developmentTasks: [],
          doneTasks: [],
        }
      })
    },
    deleteProject: (state, action: PayloadAction<{id: number}>) => {
      const index = state.projects.findIndex((p: Project) => p.id === +action.payload.id)
      if (index > -1) {
        state.projects.splice(index, 1)
      }
    },
    setCurrentProjectId: (state, action: PayloadAction<{projectId: string | undefined}>) => {
      state.currentProjectId = Number(action.payload.projectId)
    },
    changeTaskStatus: (state, action: PayloadAction<{
      currentStatus: string
      id: number
      nextStatus?: string
    }>) => {
      state.projects.map((p: Project) => {
        if (p.id === state.currentProjectId) {
          const element = removeTaskFromCurrentStatus(p.tasks, action.payload.currentStatus, action.payload.id)
          pushTaskToNextStatus(p.tasks, element, action.payload.nextStatus)
        }
      })
    },
    createNewTask: (state, action: PayloadAction<{
      status: string
      text: string

    }>) => {
      let statusTask = selectSection(action.payload.status)
      let element: Task = {
        id: Date.now(),
        text: action.payload.text || 'Update me',
        description: '',
        priority: 'low',
        subTasks: []
      }
      state.projects.map((p: Project) => {
        if (p.id === state.currentProjectId) {
          p.tasks[statusTask].push(element)
        }
      })
    },
    deleteTask: (state, action: PayloadAction<{
      section: string
      taskId: number
    }>) => {
      state.projects.map((p: Project) => {
        if (p.id === state.currentProjectId) {
          let section = selectSection(action.payload.section)
          let index = p.tasks[section].findIndex((t: Task) => t.id === action.payload.taskId)
          if (index > -1) {
            p.tasks[section].splice(index, 1)
          }
        }
      })
    },
    changeTask: (state, action: PayloadAction<{
      section: string
      taskId: number
      newText: string
      newDescription: string
      newPriority: "low" | "middle" | "high"
      newSubTasks: Array<Subtask>
    }>) => {
      state.projects.map((p: Project) => {
        if (p.id === state.currentProjectId) {
          let section = selectSection(action.payload.section)
          p.tasks[section].map((t: Task) => {
            if (t.id == action.payload.taskId) {
              t.text = action.payload.newText
              t.description = action.payload.newDescription
              t.priority = action.payload.newPriority
              t.subTasks = [...action.payload.newSubTasks]
            }
          })
        }
      })
    }
  },
})

export const {
  createProject,
  deleteProject,
  setCurrentProjectId,
  changeTaskStatus,
  createNewTask,
  deleteTask,
  changeTask,
} = projectsSlice.actions

export default projectsSlice.reducer


const pushTaskToNextStatus = (state: Tasks, element: Task, nextStatus: string | undefined): void => {
  let section = selectSection(nextStatus)
  state[section].push(element)
}

const removeTaskFromCurrentStatus = (state: Tasks, currentStatus: string, id: number): Task => {
  let section = selectSection(currentStatus)
  const index = state[section].findIndex((el: Task) => el.id === id)
  const result = state[section].splice(index, 1)[0]
  return result
}



const selectSection = (status: string | undefined): 'queueTasks' | 'developmentTasks' | 'doneTasks' => {
  switch (status) {
    case 'Queue':
      return 'queueTasks'
    case 'Development':
      return 'developmentTasks'
    case 'Done':
      return 'doneTasks'
    default:
      return 'queueTasks'
  }
}