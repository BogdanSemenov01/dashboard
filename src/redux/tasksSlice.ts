import {createSlice} from '@reduxjs/toolkit'
import type {PayloadAction} from '@reduxjs/toolkit'

const initialState:any = {
  tasks: {

  }
}

const tasksSlice = createSlice({
  name: 'tasks',
  initialState,
  reducers: {
    loadTasks: (state, action) => {
      state.tasks.push(action.payload.tasks)
    },
    changeTaskStatus: (state, action) => {
      if (state.tasks) {

        state.tasks.map((t:any) => {
          if (t.id === action.payload.id) {
            t.status = action.payload.status
          }
        })
      }
    }
  },
})


