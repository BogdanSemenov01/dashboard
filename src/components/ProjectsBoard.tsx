import React from 'react'
import { RootState, useAppDispatch, useAppSelector } from '../redux/store'
import { Project, createProject, deleteProject } from '../redux/projectsSlice'
import styled from 'styled-components';
import { Link } from 'react-router-dom'
import DeleteButton from './common/DeleteButton'
import AddButton from './common/AddButton'

const StyledProject = styled.div`
  border-radius: 3px;
  font-size: 20px;  
  padding: 0 5px;
  width: 270px;
  height: 30px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  cursor: pointer;
  background-color: white;
  color: rgba(218, 78, 101, 0.705);
  &>a {
    color: rgba(218, 78, 101, 0.705);
    text-decoration: none;
  }
  &>button{
    color: rgb(218, 37, 67);
    height: 19px;
    background: none;
    border: 1px solid rgba(218, 78, 101, 0.705);
    border-radius: 5px;
    &:hover {
      background: rgba(218, 78, 101, 0.705);
    }
  }
`;

const StyledProjectsBoard = styled.div`
  display: flex;
  flex-direction: column;
  gap: 3px;
  height: 60vh;
  background-color: rgba(218, 78, 101, 0.705);
  border: 2px solid rgb(218, 37, 67);
  padding: 10px;
  border-radius: 5px;
`;

const StyledProjectsBoardTitle = styled.div`
  color: rgb(218, 37, 67);
  font-weight: 500;
  text-align: center;
  font-size: 35px;
`;

const theme = {
  background: 'rgba(218, 78, 101, 0.705)',
  title: 'rgb(218, 37, 67)'
}

const ProjectsBoard = (): React.ReactElement => {

  const projects = useAppSelector((state: RootState) => state.projects.projects)
  const dispatch = useAppDispatch()

  const onClickDeleteProject = (event: any) => {
    dispatch(deleteProject({id: event.target.id}))
  }

  return (
    <div>
        <StyledProjectsBoard>
        <StyledProjectsBoardTitle>Projects</StyledProjectsBoardTitle>
        {projects.map((p: Project) => {
          return <StyledProject key={p.id}>
            <Link to={'/' + p.id}>
              {p.title}
            </Link>
            <DeleteButton onClick={onClickDeleteProject} id={p.id.toFixed()} />
            </StyledProject>
        })}
      <AddButton callback={createProject} theme={theme}/>
        </StyledProjectsBoard>
    </div>
  )
}

export default ProjectsBoard