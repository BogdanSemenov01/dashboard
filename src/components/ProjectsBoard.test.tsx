import { render, screen } from "@testing-library/react"
import ProjectsBoard from "./ProjectsBoard"
import * as appHooks from "../redux/store"
import { Project } from "../redux/projectsSlice"
import { MemoryRouter, Route, Routes, useLocation } from "react-router-dom"
import userEvent from "@testing-library/user-event"
const mockedDispatch = jest.spyOn(appHooks, "useAppDispatch")
const mockedSelector = jest.spyOn(appHooks, "useAppSelector")
const dispatch = jest.fn()

const projects: Array<Project> = [
  {
    title: "DashBoard",
    id: 1,
    tasks: {
      queueTasks: [],
      developmentTasks: [],
      doneTasks: [],
    },
  },
]

export const LocationDisplay = () => {
  const location = useLocation()

  return <div data-testid="location-display">{location.pathname}</div>
}

describe("Projects Board", () => {
  it("should render projects board component", () => {
    mockedSelector.mockReturnValue(projects)
    const component = render(
      <MemoryRouter initialEntries={['/']}>
        <ProjectsBoard />
      </MemoryRouter>
    )
    expect(component).toMatchSnapshot()
  })
  it("should change location after click", () => {
    mockedSelector.mockReturnValue(projects)
    render(
      <MemoryRouter initialEntries={['/']}>
        <ProjectsBoard />
        <LocationDisplay />
      </MemoryRouter>
    )
    const link = screen.getByText('DashBoard')
    userEvent.click(link)
    expect(screen.getByTestId('location-display').innerHTML).toEqual('/1')
  })
})
