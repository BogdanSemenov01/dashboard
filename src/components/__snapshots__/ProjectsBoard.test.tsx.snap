// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects Board should render projects board component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="sc-eDvSVe ia-DZiP"
        >
          <div
            class="sc-jSUZER iflYaU"
          >
            Projects
          </div>
          <div
            class="sc-hLBbgP jNGPUm"
          >
            <a
              href="/1"
            >
              DashBoard
            </a>
            <button
              class="sc-bcXHqe iPtRTz"
              id="1"
            >
              x
            </button>
          </div>
          <button
            class="sc-gswNZR fQtaQt"
          >
            +
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="sc-eDvSVe ia-DZiP"
      >
        <div
          class="sc-jSUZER iflYaU"
        >
          Projects
        </div>
        <div
          class="sc-hLBbgP jNGPUm"
        >
          <a
            href="/1"
          >
            DashBoard
          </a>
          <button
            class="sc-bcXHqe iPtRTz"
            id="1"
          >
            x
          </button>
        </div>
        <button
          class="sc-gswNZR fQtaQt"
        >
          +
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
